### ISSUE TITLE
ES Module for browsers

### ISSUE THREAD

**Issue by mlebrasseur** [buefy/buefy#1662]
### Description

It would be cool if Buefy had an [ESM (ECMAScript Module) Build for browsers](https://jakearchibald.com/2017/es-modules-in-browsers/) like Vuejs, Vue-i18n and more. For the moment I only use the Bulma CSS and would love to use Buefy instead!

Example of some ESM.browser builds:
https://cdn.jsdelivr.net/npm/vue@2.6.0/dist/vue.esm.browser.js
https://unpkg.com/browse/vue-i18n@8.14.0/dist/vue-i18n.esm.browser.min.js
https://unpkg.com/vue-router@3.0.7/dist/vue-router.esm.browser.min.js
https://unpkg.com/vuex@3.1.1/dist/vuex.esm.browser.min.js

### Why Buefy need this feature

It will be a plus if you guys release a [ESM for browsers](https://jakearchibald.com/2017/es-modules-in-browsers/) build! Because people can use ES modules natively without any build process. So, for small and medium projects/applications, it's awesome! Everything is simplier and straightfoward. ❤

**Comment by mlebrasseur** [comment]
Still not working. I got the following error:
`TypeError: Error resolving module specifier: vue`

This is not an ESM module for browsers. It still try to import vue from vue.

**Comment by jtommy** [comment]
What file are you importing?

**Comment by mlebrasseur** [comment]
I'm only importing Vue and then Buefy.
`// Imports
import Vue from "./assets/dependencies/vue/js/vue.js";
import Buefy from "./assets/dependencies/buefy/js/buefy.js";
Vue.use(Buefy);`
There's not build process in this project.
ESM for browsers are supposed to be supported natively without any build process.
😊

**Comment by jtommy** [comment]
But the correct file is buefy.esm.js

**Comment by mlebrasseur** [comment]
I renamed the files. So i'm using: 
Vue: https://unpkg.com/vue@2.6.10/dist/vue.esm.browser.min.js
Buefy: https://unpkg.com/buefy@0.8.3/dist/buefy.esm.min.js

**Comment by mlebrasseur** [comment]
Just looking at the code inside the Buefy ESM file, the first line is incorrect for an ESM for browsers.
`import Vue from 'vue';`

If we look at the official ESM build for browsers released by Vue:
https://cdn.jsdelivr.net/npm/vue@2.6.0/dist/vue.esm.browser.js
https://unpkg.com/browse/vue-i18n@8.14.0/dist/vue-i18n.esm.browser.min.js
https://unpkg.com/vue-router@3.0.7/dist/vue-router.esm.browser.min.js
https://unpkg.com/vuex@3.1.1/dist/vuex.esm.browser.min.js

There's no such import of Vue.

**Comment by afwn90cj93201nixr2e1re** [comment]
@mlebrasseur give an example of page which use this importing style asap.

**Comment by afwn90cj93201nixr2e1re** [comment]
@jtommy it's not bug, can be closed, this import is expected. Here no workaround for that. The only one just remove this line after build.

**Comment by mlebrasseur** [comment]
@afwn90cj93201nixr2e1re This is a "feature request", look at the label.

**Comment by afwn90cj93201nixr2e1re** [comment]
It's not. It's already done. But you wanna something another, so, you should create window.Vue and remove import vue from vue for yourself.
It's how ESM works, it's expected.

**Comment by mlebrasseur** [comment]
@afwn90cj93201nixr2e1re I think you don't understand what a ES Module FOR BROWSER is. If you think i'm importing Vue this way: "import vue from vue" you're wrong because how can you import vue from vue if you don't have build steps. ESM for browsers are supported natively on the browsers directly. You can read more about this here: https://vuejsdevelopers.com/2019/02/04/vue-es-module-browser-build/ and here https://jakearchibald.com/2017/es-modules-in-browsers/

**Comment by afwn90cj93201nixr2e1re** [comment]
vuex, vue router, vue i18n not frameworks, it's modules, which are not related to Vue at all. They are not using Vue. They get vue thru install, buefy also do this stuff, but there too much additional stuff in core which are using vue.

**Comment by ronreiter** [comment]
I actually need this right now as well. Funny to see such a relevant discussion going on as we speak.
I also want to write a lightweight dashboard example of using Buefy and the import "vue" is not making this possible.
the only question is - if Buefy requires Vue then how should it refer to Vue? as a global window variable (e.g. just remove the import) or from a specific CDN provider using an absolute link?

**Comment by mlebrasseur** [comment]
@afwn90cj93201nixr2e1re I get your point. Thank you.
@jtommy What do you think about it? It's really not possible to make a ESM for browser build of Buefy?

**Comment by afwn90cj93201nixr2e1re** [comment]
> as a global window variable (e.g. just remove the import)
Yes, like that. That the only one right solution which is used now in vuex, vue-router and other stuff.
For more info you can check them esm builds, if Vue is not exists in install method, then they get it from window object.

@mlebrasseur you r crazy rly, what do you dont get from our answers? It's how rollup compiler works, and that's works as expected, as in specifications. You need browser build, not esm.
Caused by this line: https://github.com/buefy/buefy/blob/dev/rollup.config.js#L145, which is not exists in browser build. We ended up with that, and there no clear solutions for skipping import inside esm file.
![изображение](https://user-images.githubusercontent.com/9306996/70351536-97cf9400-1882-11ea-977e-ccf4e78e03b1.png)
that's how bootstrap works.

**Comment by afwn90cj93201nixr2e1re** [comment]
@ronreiter i've invistigate it few days ago, and that's why import is still in file, so, for get rid of import the only one thing that you need - existing window.Vue object and that's all, then you can delete import from top and end up with it.

![изображение](https://user-images.githubusercontent.com/9306996/70351910-63a8a300-1883-11ea-85ad-9803c41a0253.png)
![изображение](https://user-images.githubusercontent.com/9306996/70351933-7327ec00-1883-11ea-8c66-7ee74b4526d3.png)
![изображение](https://user-images.githubusercontent.com/9306996/70351948-7b802700-1883-11ea-980a-833488e23bf7.png)
![изображение](https://user-images.githubusercontent.com/9306996/70351955-8044db00-1883-11ea-97e8-95162425184a.png)
![изображение](https://user-images.githubusercontent.com/9306996/70351975-8cc93380-1883-11ea-89b6-a497b2cd19e5.png)
![изображение](https://user-images.githubusercontent.com/9306996/70351997-96529b80-1883-11ea-9485-882163416f19.png)

**Comment by afwn90cj93201nixr2e1re** [comment]
Or just remove all of this stuff
https://github.com/buefy/buefy/blob/dev/src/components/snackbar/index.js#L1.
I can easy fix it right now, but not sure about jtommy reactivity.

**Comment by ronreiter** [comment]
@jtommy WDYT?

**Comment by jtommy** [comment]
@ronreiter It's no easy, Rollup doesn't allow to skip Vue import. I remember to tried but without a good solution

**Comment by afwn90cj93201nixr2e1re** [comment]
It's easy, setting temp env on esm build and put code inside if, that the only one right solution, but i think that they should use browser build btw.
Emm, but also we can just create global variable like config, and bind Vue instance on install and remove useless code https://github.com/buefy/buefy/blob/dev/src/components/snackbar/index.js#L27, we already have everything for easy fix.

**Comment by afwn90cj93201nixr2e1re** [comment]
![изображение](https://user-images.githubusercontent.com/9306996/70378272-cbb8c100-1937-11ea-9b34-0c7c04db44a6.png)


And that's what we gonna get at end.

**Comment by afwn90cj93201nixr2e1re** [comment]
See #2063 for fix.

**Comment by afwn90cj93201nixr2e1re** [comment]
Dist:

[without_import_buefy.esm.zip](https://github.com/buefy/buefy/files/3935385/without_import_buefy.esm.zip)

**Comment by ronreiter** [comment]
I'd love to have this published and available on a CDN.
Can I assist somehow? I'd say this would have to wait for the next release, right?

**Comment by afwn90cj93201nixr2e1re** [comment]
Jtommy should merge it, then you can get it directly from registry.npm.com/buefy...

Without merging we can't do anything.

**Comment by jtommy** [comment]
@ronreiter did you test @afwn90cj93201nixr2e1re build?

**Comment by mlebrasseur** [comment]
> @mlebrasseur you r crazy rly, what do you dont get from our answers? It's how rollup compiler works, and that's works as expected, as in specifications. You need browser build, not esm.
Caused by this line: https://github.com/buefy/buefy/blob/dev/rollup.config.js#L145, which is not exists in browser build. We ended up with that, and there no clear solutions for skipping import inside esm file.
@afwn90cj93201nixr2e1re I'm just curious if that's possible to have a esm for browser cdn for Buefy. I've never used Rollup. So that's why I created this feature request in the first place.
Sounds like you found a working solution so don't call me crazy lol. 👀
